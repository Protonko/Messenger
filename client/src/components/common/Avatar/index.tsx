import {useMemo, FC} from 'react'
import classNames from 'classnames'
import {Sizes} from 'models/common/sizes'
import {ColorGeneratorByName} from 'utils/ColorGeneratorByName'
import {isDarkColor} from 'utils/isDarkColor'
import {Loader} from 'components/common/Loader'

export interface IPropsAvatar {
  src: string
  name?: string
  size?: Sizes
  customStyles?: string
}

export const Avatar: FC<IPropsAvatar> = ({
  src,
  name,
  size = Sizes.MEDIUM,
  customStyles,
}) => {
  const backgroundColor = useMemo(
    () => new ColorGeneratorByName(name ?? '').generate(),
    [src, name],
  )
  const avatarClassname = classNames(
    'avatar',
    {'avatar--sm': size === Sizes.SMALL},
    {'avatar--lg': size === Sizes.LARGE},
    {[customStyles ?? '']: customStyles},
  )

  if (!src && name) {
    const classNamesText = classNames(
      'avatar__text',
      {'avatar__text--light': isDarkColor(backgroundColor)},
      {'avatar__text--dark': !isDarkColor(backgroundColor)},
    )

    return (
      <div className={avatarClassname} style={{backgroundColor}}>
        <span className={classNamesText}>{name.charAt(0)}</span>
      </div>
    )
  }

  return (
    <div className={avatarClassname}>
      <img className="avatar__image" src={src} alt="" />
      <Loader />
    </div>
  )
}
