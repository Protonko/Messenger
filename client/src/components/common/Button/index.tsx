import {FC, HTMLProps, ReactNode} from 'react'

interface IButtonProps extends HTMLProps<HTMLButtonElement> {
  text?: string
  icon?: ReactNode
  leftAddons?: ReactNode
  rightAddons?: ReactNode
  type?: 'button' | 'submit' | 'reset'
}

export const Button: FC<IButtonProps> = ({
  text,
  icon,
  leftAddons,
  rightAddons,
  ...buttonProps
}) => {
  const renderButtonContent = () => {
    return (
      <>
        {text && (
          <span className="button__text">
            {text}
          </span>
        )}
        {icon && (
          <span className="button__icon">
            {icon}
          </span>
        )}
      </>
    )
  }

  return (
    <button className="button" {...buttonProps}>
      {leftAddons}
      {renderButtonContent()}
      {rightAddons}
    </button>
  )
}
