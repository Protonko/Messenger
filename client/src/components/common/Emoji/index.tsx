import 'emoji-mart/css/emoji-mart.css'
import {FC, useState, useRef} from 'react'
import {EmojiData, Picker} from 'emoji-mart'
import {ReactComponent as Smile} from 'assets/icons/smile.svg'
import {useOutsideClick} from 'hooks/useOutsideClick'

export interface IPropsEmoji {
  onSelect: (emoji: EmojiData) => void
}

export const Emoji: FC<IPropsEmoji> = ({onSelect}) => {
  const [open, setOpen] = useState(false)
  const ref = useRef(null)
  const initiatorRef = useRef(null)

  useOutsideClick(ref, () => setOpen(false), initiatorRef)

  const onSelectEmoji = (emoji: EmojiData) => {
    onSelect(emoji)
    setOpen(false)
  }

  return (
    <div className="emoji">
      {open && (
        <div className="emoji__popup" ref={ref}>
          <Picker native={true} onSelect={onSelectEmoji} />
        </div>
      )}

      <Smile
        className="emoji__button"
        ref={initiatorRef}
        onClick={() => setOpen(!open)}
      />
    </div>
  )
}
