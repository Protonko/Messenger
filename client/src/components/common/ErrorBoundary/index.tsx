import {Component, ErrorInfo} from 'react';

interface IProps {}
interface IState {
  error: null | Error;
  errorInfo: null | ErrorInfo;
  hasError: boolean;
}

export class ErrorBoundary extends Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);

    this.state = {
      error: null,
      errorInfo: null,
      hasError: false,
    }
  }

  static getDerivedStateFromError(error: Error) {
    console.log(error)
    return {
      hasError: true,
    };
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    this.setState({
      error: error,
      errorInfo: errorInfo
    })
  }

  _renderDetailInfo() {
    if (process.env.NODE_ENV === 'development') {
      return (
        <div>

        </div>
      )
    }

    return null
  }

  render() {
    if (this.state.hasError) {
      return (
        <>
          <h1>Что-то пошло не так.</h1>

          {process.env.NODE_ENV === 'development' && (
            <details style={{ whiteSpace: 'pre-wrap' }}>
              {this.state.error && this.state.error.toString()}
              <br />
              {this.state.errorInfo?.componentStack}
            </details>
          )}
        </>
      )
    }

    return this.props.children;
  }
}
