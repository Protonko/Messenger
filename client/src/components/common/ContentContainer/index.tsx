import type {FC} from 'react'
import {Loader} from 'components/common/Loader'
import {Text} from 'components/common/Text'

interface ContentContainerProps {
  loading: boolean
  errorMessage: string | null
  children: JSX.Element
}

export const ContentContainer: FC<ContentContainerProps> = ({
  loading,
  errorMessage,
  children,
}) => {
  if (errorMessage) {
    return <Text>{errorMessage}</Text>
  }

  if (loading) {
    return <Loader center={true} />
  }

  return children
}
