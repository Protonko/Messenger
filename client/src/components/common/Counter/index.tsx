// types
import {Status} from 'models/common/status'

import {FC} from 'react'
import classNames from 'classnames'

export interface IPropsCounter {
  count: number | '99+',
  status?: Status,
}

export const Counter: FC<IPropsCounter> = ({
  count,
  status = Status.ACTIVE,
}) => {
  const classNameStatus = classNames(
    'counter',
    {'counter--muted': status === Status.MUTED}
  )

  return (
    <div className={classNameStatus}>
      <span className="counter__text">
        {count}
      </span>
    </div>
  )
}
