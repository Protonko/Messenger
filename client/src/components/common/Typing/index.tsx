import type {FC} from 'react'
import classNames from 'classnames'
import {Sizes} from 'models/common/sizes'

interface IPropsTyping {
  customStyles?: string
  size?: Sizes
}

export const Typing: FC<IPropsTyping> = ({
  customStyles,
  size = Sizes.MEDIUM,
}) => {
  const classNamesTyping = classNames(
    'typing',
    {[customStyles ?? '']: !!customStyles},
    {'typing--large': size === Sizes.LARGE},
    {'typing--small': size === Sizes.SMALL},
  )

  return (
    <div className={classNamesTyping}>
      <p className="typing__text">Typing</p>
      <div className="typing__loader">
        <div className="typing__loader-ellipse" />
        <div className="typing__loader-ellipse" />
        <div className="typing__loader-ellipse" />
        <div className="typing__loader-ellipse" />
      </div>
    </div>
  )
}
