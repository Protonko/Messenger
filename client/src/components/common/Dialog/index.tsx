import type {FC} from 'react'
import type {IDialog} from 'models/dialog'
import {ReadStatus} from 'models/common/status'
import {TextSize, TextTypes} from 'models/common/text'
import {ReactComponent as DoubleCheck} from 'assets/icons/double-check.svg'
import {ReactComponent as Check} from 'assets/icons/check.svg'
import {COLORS} from 'static/colors'
import {Text} from 'components/common/Text'
import {Counter} from 'components/common/Counter'
import {Avatar} from 'components/common/Avatar'

export const Dialog: FC<IDialog> = ({
  name,
  description,
  avatar,
  edited,
  createdAt,
  updatedAt,
  messages,
  status,
  readStatus,
}) => {
  const counter = messages.toString().length > 2 ? '99+' : messages

  const renderStatus = () => {
    if (!!counter) {
      return <Counter count={counter} status={status}/>
    }

    if (readStatus) {
      const props = {
        width: 15,
        height: 15,
        color: COLORS.dustyGray,
      }
      return readStatus === ReadStatus.READ ? <DoubleCheck {...props} /> : <Check {...props} />
    }

    return null
  }

  return (
    <div className="dialog">
      <Avatar src={avatar} />

      <div className="dialog__data">
        <div className="dialog__text">
          <Text
            type={TextTypes.h4}
            customStyles="dialog__text-title"
            numberOfLines={1}
          >
            {name}
          </Text>
          <Text>{description}</Text>
        </div>

        <div className="dialog__info">
          <Text
            type={TextTypes.mixed}
            size={TextSize.EXTRA_SMALL}
            customStyles="dialog__info-date"
          >
            {new Date(edited ? updatedAt : createdAt).toLocaleDateString()}
          </Text>

          {renderStatus()}
        </div>
      </div>
    </div>
  )
}
