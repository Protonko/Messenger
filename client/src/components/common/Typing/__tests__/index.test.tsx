import {shallow, ShallowWrapper} from 'enzyme'
import {Sizes} from 'models/common/sizes'
import {IPropsTyping, Typing} from 'components/common/Typing'

const ELEMENT_SELECTORS = {
  typing: 'typing',
  custom: 'custom',
  large: 'typing--large',
  small: 'typing--small',
}

describe('Typing', () => {
  let props: IPropsTyping
  let component: ShallowWrapper

  beforeEach(() => {
    props = {}
  })

  it('Should render Typing component', () => {
    component = shallow(<Typing {...props} />)

    expect(component.find(`.${ELEMENT_SELECTORS.typing}`).length).toBe(1)
    expect(component).toMatchSnapshot('Typing')
  })

  it('Should render Typing component with custom styles', () => {
    props.customStyles = ELEMENT_SELECTORS.custom
    component = shallow(<Typing {...props} />)

    expect(component.find(`.${ELEMENT_SELECTORS.typing}`).length).toBe(1)
    expect(component).toMatchSnapshot('Typing with custom styles')
  })

  it('Should render large Typing component', () => {
    props.size = Sizes.LARGE
    component = shallow(<Typing {...props} />)

    expect(component.find(`.${ELEMENT_SELECTORS.large}`).length).toBe(1)
    expect(component).toMatchSnapshot('Typing large')
  })

  it('Should render small Typing component', () => {
    props.size = Sizes.SMALL
    component = shallow(<Typing {...props} />)

    expect(component.find(`.${ELEMENT_SELECTORS.small}`).length).toBe(1)
    expect(component).toMatchSnapshot('Typing small')
  })
})
