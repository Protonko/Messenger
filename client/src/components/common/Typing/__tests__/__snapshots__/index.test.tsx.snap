// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typing Should render Typing component with custom styles: Typing with custom styles 1`] = `
<div
  className="typing custom"
>
  <p
    className="typing__text"
  >
    Typing
  </p>
  <div
    className="typing__loader"
  >
    <div
      className="typing__loader-ellipse"
    />
    <div
      className="typing__loader-ellipse"
    />
    <div
      className="typing__loader-ellipse"
    />
    <div
      className="typing__loader-ellipse"
    />
  </div>
</div>
`;

exports[`Typing Should render Typing component: Typing 1`] = `
<div
  className="typing"
>
  <p
    className="typing__text"
  >
    Typing
  </p>
  <div
    className="typing__loader"
  >
    <div
      className="typing__loader-ellipse"
    />
    <div
      className="typing__loader-ellipse"
    />
    <div
      className="typing__loader-ellipse"
    />
    <div
      className="typing__loader-ellipse"
    />
  </div>
</div>
`;

exports[`Typing Should render large Typing component: Typing large 1`] = `
<div
  className="typing typing--large"
>
  <p
    className="typing__text"
  >
    Typing
  </p>
  <div
    className="typing__loader"
  >
    <div
      className="typing__loader-ellipse"
    />
    <div
      className="typing__loader-ellipse"
    />
    <div
      className="typing__loader-ellipse"
    />
    <div
      className="typing__loader-ellipse"
    />
  </div>
</div>
`;

exports[`Typing Should render small Typing component: Typing small 1`] = `
<div
  className="typing typing--small"
>
  <p
    className="typing__text"
  >
    Typing
  </p>
  <div
    className="typing__loader"
  >
    <div
      className="typing__loader-ellipse"
    />
    <div
      className="typing__loader-ellipse"
    />
    <div
      className="typing__loader-ellipse"
    />
    <div
      className="typing__loader-ellipse"
    />
  </div>
</div>
`;
